libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1650,3 +1671,216 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
 	solAssert(_expression.annotation().type, "Type of expression not set.");
 	return *_expression.annotation().type;
 }
+
+bool IRGeneratorForStatements::visit(TryStatement const& _tryStatement)
+{
+	solAssert(!m_returnInfo.has_value(), "");
+
+	m_returnInfo.reset();
+	_tryStatement.externalCall().accept(*this);
+	solAssert(m_returnInfo.has_value(), "");
+
+	auto const trySuccessCondition = m_context.trySuccessConditionVariable(_tryStatement.externalCall());
+
+	m_code << "switch iszero(" << trySuccessCondition << ")\n";
+
+	m_code << "case 0 { // success case\n";
+	TryCatchClause const& successClause = *_tryStatement.clauses().front();
+	decodeTryCallReturnParameters(successClause, _tryStatement.externalCall().annotation().type);
+	m_returnInfo.reset();
+	successClause.block().accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement.clauses());
+	m_code << "}\n";
+
+	return false;
+}
+
+string IRGeneratorForStatements::decodeReturnParameters()
+{
+	ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+
+	return Whiskers(R"(
+		<?dynamicReturnSize>
+			// copy dynamic return data out
+			returndatacopy(<pos>, 0, returndatasize())
+		</dynamicReturnSize>
+
+		// update freeMemoryPointer according to dynamic return size
+		mstore(<freeMemoryPointer>, add(<pos>, and(add(<returnSize>, 0x1f), not(0x1f))))
+
+		// decode return parameters from external try-call into <retVars>
+		let <retVars> := <abiDecode>(<pos>, add(<pos>, <returnSize>))
+	)")
+	("retVars", IRVariable(m_returnInfo->functionCall).commaSeparatedList())
+	("pos", m_returnInfo->returndataVariable)
+	("abiDecode", abi.tupleDecoder(m_returnInfo->returnTypes, true))
+	("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+	("dynamicReturnSize", m_returnInfo->dynamicReturnSize)
+	("returnSize",
+		m_context.evmVersion().supportsReturndata()
+			? "returndatasize()"s
+			: to_string(m_returnInfo->estimatedReturnSize)
+	).render();
+}
+
+void IRGeneratorForStatements::decodeTryCallReturnParameters(TryCatchClause const& _successClause, TypePointer _callReturnType)
+{
+	solAssert(m_returnInfo.has_value(), "");
+	verifyTrySuccessClause(_successClause, _callReturnType);
+	if (!_successClause.parameters())
+		return;
+
+	m_code << decodeReturnParameters();
+
+	size_t i = 0;
+	for (ASTPointer<VariableDeclaration> const& varDecl: _successClause.parameters()->parameters())
+	{
+		solAssert(varDecl, "");
+		define(
+			m_context.addLocalVariable(*varDecl),
+			IRVariable(m_returnInfo->functionCall).tupleComponent(i++)
```

Comment:
Do you have a test where the function call returns exactly one stack component? I.e. where the type of ``m_returnInfo->functionCall`` is not a tuple? Because for that this will break...

Reply:
Now I'm having one and fixed the code path for that.